### ISSUE TITLE
Native CSS support

### ISSUE THREAD

**Issue by sokra** [webpack/webpack#14893]
Current State: opt-in via `experiments.css: true`

Explainer (not everything is implemented yet):

* `experiments.css: true` enables the native css support in webpack
* This will be enabled by default in webpack 6
* There will be multiple modes depending on how you reference CSS
  * A: `import "./style.css"`: Attaches the stylesheet to the document as side effect of importing. `@import` and `url()` are resolved.
  * B: `import stylesheet from "./style.css" asset { type: "css" }`: Exports the unattached stylesheet as default export. You can adopt it according to the spec. `@import` is not allowed (yet), but `url()` will be resolved.
  * C: `import { main } from "./style.module.css"`: Like A, but renames all css classes, keyframes and css variables and exports the new names. Global selectors are only allowed with explicit `:global` selector prefix. `@import` and `url()` are resolved.
  * D: `new URL("./style.css", import.meta.url)`: Gives an URL to a stylesheet including all styles. `@import` and `url()` are resolved.
  * E: `import { main } from "./style.module.css"`: Like C, but for the node.js target. It generates a JSON module with the exports from the CSS.
* A and C will bundle all CSS of a chunk into a single CSS file (per chunk)
* splitChunks will allow to move CSS around
* `output.cssFilename` and `output.cssChunkFilename` allows to configure the filename template for css files. By default it copies `filename` resp. `chunkFilename` with extension changed to `.css`
* HMR will update CSS when you edit it.
* There is external CSS that will lead to `@import "..."` in the output css files
* There are external assets that will lead to a external url.
* class names are based on module ids, so you need to make sure that SSR and Client module ids match
  * E1: Compute module ids in a deterministic way -> DeterministicModuleIdsPlugin
  * E2: Store module ids in a file and load them on the other side -> SyncModuleIdsPlugin
* In addition to the default bindings, you can use `type: "css/..."` in rules to apply loader results as css. e. g. for Sass support.
* This will replace `mini-css-extract-plugin` and `css-loader`
* It will be more performant, by using a css tokenizer instead of postcss
* There is a css minimizer in production mode.

Implemented:

* A
* `@import url("...");`
* `@import "...";`
* `url()`
* chunking
* on demand loading
* HMR
* splitChunks
* `output.cssFilename`
* external type `css-import`
* external type `asset`
* `:export` blocks
* C (partially)
  * classes
  * ids
  * keyframes + animation
  * css variables + var()
  * `var(xxx from "...")`
  * `var(xxx from global)` 
* correct css order in output file
* warnings for ordering issues
* E
* E1
* E2
* `image-set()` 
* `@import url("...") layer(test);`
* `@import url("...") supports(...);`
* `@import url("...") media query;`
* move all external imports to top
* do not ignore `#import` and make them external
* inherit layer, supports(...) and media from parent module
* supports `style` resolution in package.json and test it + test support loader usage
* Implement CSS Module `fetchpriority`
* prefetch/preload for CSS
* external import with media/supports
* Correct `publicPath` for assets modules inside CSS files
* pathinfo for `@import`
* prefeth/preload only for CSS `import()` generates prefetch/preload runtime for JS
* default `type: css` enables CSS modules support (i.e. for CSS without CSS modules you need set `type: "css/global"`), which might be a bit confusing for developers, because some developer just want to use pure CSS and parsing CSS modules can affect on perf, I think we should do:
  - `type: "css/auto"` - enable/disable CSS modules based on path, it is `css/module` or `css`
  - `type: "css"` - no CSS modules and don't try to parse CSS modules things
  - `type: "css/module"` - CSS modules with `local` mode
  - `type: "css/global"` CSS modules with `global` mode
* rewrite to be align with CSS spec 
* `src()`
* unstable `?hash` generation in --webpack-main when you have multiple same modules
* improve logic in callback (more perfomance?) + better error reporting for `@import`
* implement `/* webpackIgnore: true */` for `url()`/`image-set()`
* `url()` in css leaves an asset JS Module in js chunk
* `:import` blocks
*  `@value`
* https://github.com/webpack/webpack/issues/17611 - allow link be before/after script + bug with safari 
* * support legacy browser? https://github.com/webpack/webpack/issues/16147

Not implemented/tested, but Planned (ordered by priority):
 
* C (partially)
  * `composes: xxx`
  * `composes: xxx from "..."`
  * `composes: xxx from global`
* re-export for `:export`/`@value`/`var()`/classes and nested `@import`
* icss interpolation for everything
* `local()` and `global()` in declarations 
* more at-rules rename in css modules
* move tests from`postcss-modules-*` packages 
* `css/global` test
* move tests from css-loader
* moove test from extract plugin
* internal refactor CSS parser + use `ICSS` prefix for deps + inheritance for deps
* D
* B
* remove empty async chunks
* CSS entry without JS file (option to do only extract without runtime + for dynamic?) and hmr
* import sheet from './styles.css' assert { type: 'css' };
* `exportGlobals`
* improve hashing of locals for production
* `pure` mode 
* `image()` 
* css minimizer 
* contenthashing for CSS files
* normalize media/supports/layer to lowercase?
* `url("font.svg#svgFontName")` created invalid filename 
* wasm css tokenizer

**Comment by alexander-akait** [comment]
bump

**Comment by alexander-akait** [comment]
Broken URL - https://github.com/webpack/webpack/issues/16969#issue-1664483188

**Comment by hardfist** [comment]
@alexander-akait  I'm wondering how experiments.css working with the current css-loader ecosystem,for example https://github.com/seek-oss/css-modules-typescript-loader which consume the result of css-loader, but since experiments.css handle css after loader there're no chances for other loader to get css-module intermediate result, will experiments.css expose some hooks for this scenarios?

**Comment by alexander-akait** [comment]
The main idea of the new CSS pipeline - avoid generate JS and CSS multiple times, like we have right now

**Comment by hardfist** [comment]
i do agree this should be implemented as plugin and no need to be compatible，how this plugin should be  implemented becomes a problem，should this plugin consume css dep directly and do the same codegen logic as internal css codegen or should it just consume the intermediate result of intetnal css module processing

**Comment by alexander-akait** [comment]
@hardfist 

> i do agree this should be implemented as plugin and no need to be compatible，how this plugin should be implemented becomes a problem，should this plugin consume css dep directly and do the same codegen logic as internal css codegen or should it just consume the intermediate result of intetnal css module processing

Yeah, we need to think about it, maybe we can introduce extra hooks to simplify this process, let's keep this here, We wanted to finish normal CSS support firslty and then I will think about it and put in our test cases a simplified example

**Comment by hardfist** [comment]
I'm not sure whether all the options in  css-loader and mini-css-extract-plugin will be supported in experiments.css, and we're migrating some applications from css-loader & mini-css-extract-plugin to experiments.css and missing some functionalities, so maybe I can track these in this issue or file separate issue?
* https://webpack.js.org/loaders/css-loader/#sourcemap user want to ignore css sourcemap but keep js sourcemap, so an separate devtool option for css may be needed
if these functionalities are accepted we're willing to implement it for webpack

**Comment by fregante** [comment]
It's unclear how to get the old `extract` + `css-loader` setup: generate CSS files **and do nothing else.**

Currently it appears that `experiments.css: true` generates the CSS files but also tries to load them, causing this error in my environment (web extension)

```
Uncaught TypeError: Failed to execute 'getComputedStyle' on 'Window': parameter 1 is not of type 'Element'.
    at loadCssChunkData (refined-github.js:28644:21)
```


Which wasn't the case with the following configuration:


```js
module: {
	rules: [
		{
			test: /\.css$/,
			use: [
				MiniCssExtractPlugin.loader,
				'css-loader',
			],
		},
	]
],
plugins: {
	new MiniCssExtractPlugin(),
},
```

**Comment by jeromehan** [comment]
> getComputedStyle

This might help you https://github.com/webpack/webpack/issues/16147#issuecomment-1583612714

**Comment by fregante** [comment]
No, webpack should not run any JavaScript if I ask it to generate a CSS file, so that isn't useful for this scenario.

**Comment by User-DK** [comment]
Hello everyone,

I'm interested in contributing to the issue of Native CSS support for GSOC 2024, particularly since Webpack is participating in this year's program. Could you please guide me on where to find the most important details of this issue? I'm eager to gather the necessary information to tailor my ideas and proposed solutions effectively.

Thank you!

**Comment by fregante** [comment]
Inactive but not done yet

**Comment by francescocaveglia** [comment]
Hi,

there is any way to opt-out from the css variables rewrite?

Another thing, is possible to ignore the `url()`s, like with the [url](https://github.com/webpack-contrib/css-loader?tab=readme-ov-file#url) option of css-loader?

**Comment by alexander-akait** [comment]
@francescocaveglia 

> there is any way to opt-out from the css variables rewrite and use localIdentName only for css classes?

can you provide why you need disable it?

> Another thing, is possible to ignore the url()s, like with the [url](https://github.com/webpack-contrib/css-loader?tab=readme-ov-file#url) option of css-loader?

https://webpack.js.org/plugins/ignore-plugin/

**Comment by francescocaveglia** [comment]
Hi @alexander-akait 

> can you provide why you need disable it?

In our project we use an external web components library and we have many customizations made through CSS variables.
Having them rewritten means we should wrap every instance with `:global`

> https://webpack.js.org/plugins/ignore-plugin/

It is possible to configure it to ignore "every resource inside a `url()` call"?

Thanks

**Comment by alexander-akait** [comment]
> In our project we use an external web components library and we have many customizations made through CSS variables.
> Having them rewritten means we should wrap every instance with :global

Do you have `:local`? We can make it an option for this

> It is possible to configure it to ignore "every resource inside a url() call"?

You can use this as function, what setup do you have for `css-loader`?

**Comment by fregante** [comment]
Could this support request be moved out of here? It’s off topic

**Comment by intellild** [comment]
I’m a bit concerned about the broader implications of treating `new URL("style.css", import.meta.url)` as a bundled CSS dependency instead of a copied asset. If CSS receives this special handling, should similar references like `new URL("./script.ts", import.meta.url)` also become bundled entry/dependency references in the future, rather than asset URLs? I’m also wondering how this behavior should relate to `require.resolve`, and eventually `import.meta.resolve`, since all of these APIs can be used to obtain resolved resource URLs or paths. It would be helpful to clarify where webpack draws the line between “resolve to an asset URL” and “include this file in the module graph and process it according to its type,” so users can rely on consistent behavior going forward.

**Comment by alexander-akait** [comment]
>  If CSS receives this special handling, should similar references like new URL("./script.ts", import.meta.url) also become bundled entry/dependency references in the future, rather than asset URLs?

Yes, we want to do it, for HTML too

>  It would be helpful to clarify where webpack draws the line between “resolve to an asset URL” and “include this file in the module graph and process it according to its type,” so users can rely on consistent behavior going forward.

Good question, we are still think about it, should be do it by default or provide an option(s) for parsers to enable it or not, also maybe we will implement special webpack comments to control it inside code, regarding URL construction - we make it only when we can resolve URL, if we can't we keep code as is

We are open to feedback about default behavior

Maybe you can provide cases where you don't want to include modules in the graph using new URL?

**Comment by intellild** [comment]
> > If CSS receives this special handling, should similar references like new URL("./script.ts", import.meta.url) also become bundled entry/dependency references in the future, rather than asset URLs?
> 
> Yes, we want to do it, for HTML too
> 
> > It would be helpful to clarify where webpack draws the line between “resolve to an asset URL” and “include this file in the module graph and process it according to its type,” so users can rely on consistent behavior going forward.
> 
> Good question, we are still think about it, should be do it by default or provide an option(s) for parsers to enable it or not, also maybe we will implement special webpack comments to control it inside code, regarding URL construction - we make it only when we can resolve URL, if we can't we keep code as is
> 
> We are open to feedback about default behavior
> 
> Maybe you can provide cases where you don't want to include modules in the graph using new URL?

I don’t currently have a concrete use case for excluding modules from the graph when using `new URL`.

I do have some examples where I’d like the modules to be included in the graph, such as https://github.com/intellild/resolve-examples . That said, I’m not sure whether `new URL` is the best fit here, or whether we should consider alternatives like `require.resolve` or `import.meta.resolve`.

The challenge is that users may expect `require.resolve("./foo.ts")` to be transformed into `require.resolve("./foo.js")`, allowing the actual path to be resolved at runtime.

**Comment by intellild** [comment]
I think assets can be handled through module rules, so that part may not be an issue.

The main question now is whether we should use `new URL`, `require.resolve`, or `import.meta.resolve`.

**Comment by alexander-akait** [comment]
We don't touch `import.meta.resolve`, in future we should allow to bundle it, but by default it will be false, only when I set it to true for module or global for all modules, `new URL` is resolving and including in graph only when we can really resolve it, otherwise it will be as is in code, `require.resolve` - we handle it by default due historical reasons

**Comment by intellild** [comment]
🤔 Maybe we can just keep the relative path in the output,  `new URL("./script.ts", import.meta.url)` becomes `new URL("./output-chunk.js", import.meta.url)`.

Then in browser or node, it is resolved to real url path.
For lib authors, it can still be consumed by downstream bundlers.

**Comment by alexander-akait** [comment]
@intellild generation of analyzable ECMA modules output is included in our plans for the next minor release

**Comment by nathanielprice713-coder** [comment]
That’s great news about the plans for the next minor release. Having
analyzable ECMA modules in the output will definitely help with consistency
and downstream bundling.

I'll look forward to seeing how the implementation for native CSS and HTML
progress.

**Comment by alexander-akait** [comment]
We can close it, css supported ship, only couple features left - more safe optimizations for css and html, new URL resolving, but we have PRs for them and they still need to be discussed
